### ISSUE TITLE
Unknown table identifier when trying to look up CustomTable or TableImpl whose Schema doesn't reference its tables

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11502]
This test currently fails:

```java
static class S1UpperImplicitSchema extends TableImpl/<Record> {
    static final S1UpperImplicitSchema S1_UPPER_IMPLICIT_SCHEMA = new S1UpperImplicitSchema();
    final TableField<?, Integer>       COL1                     = createField(DSL.name("COL1"), INTEGER);
    final TableField<?, String>        COL2                     = createField(DSL.name("COL2"), VARCHAR(50));
    final TableField<?, Date>          COL3                     = createField(DSL.name("COL3"), DATE);

    private S1UpperImplicitSchema() {
        super(DSL.name("UPPER"), DSL.schema("S1"));
    }
}

@Test
public void testParseWithMetaLookupsImplicitSchema() {
    withMeta(S1_UPPER_IMPLICIT_SCHEMA);

    assertSelectDataTypeEquals(INTEGER, "select S1.UPPER.COL1 from S1.UPPER");
    assertSelectDataTypeEquals(VARCHAR(50), "select UPPER.COL2 from S1.UPPER");
    assertSelectDataTypeEquals(DATE, "select COL3 from S1.UPPER");

    assertThrows(ParserException.class, () -> dsl.parser().parseSelect("select S1.UPPER.COL1 from UPPER"));
    assertThrows(ParserException.class, () -> dsl.parser().parseSelect("select S1.UPPER.COL1 from S3.UPPER"));
}
```

With this test utility, making use of  `TableMetaData`:

```java
private void withMeta(Table<?>... tables) {
    configuration.set((MetaProvider) () -> dsl.meta(tables));
    configuration.settings().setParseWithMetaLookups(THROW_ON_FAILURE);
}
```

The exception being:

```
org.jooq.impl.ParserException: Unknown table identifier: [1:35] select S1.UPPER.COL1 from S1.UPPER[*]
	at org.jooq.impl.ParserContext.exception(ParserImpl.java:13238)
	at org.jooq.impl.ParserContext.lookupTable(ParserImpl.java:13561)
	at org.jooq.impl.ParserContext.parseTableName(ParserImpl.java:11027)
	at org.jooq.impl.ParserContext.parseTableFactor(ParserImpl.java:6471)
	at org.jooq.impl.ParserContext.parseLateral(ParserImpl.java:6336)
	at org.jooq.impl.ParserContext.parseTable(ParserImpl.java:6321)
	at org.jooq.impl.ParserContext.parseTables(ParserImpl.java:6314)
	at org.jooq.impl.ParserContext.parseQueryPrimary(ParserImpl.java:1588)
	at org.jooq.impl.ParserContext.parseQueryTerm(ParserImpl.java:1445)
	at org.jooq.impl.ParserContext.parseQueryExpressionBody(ParserImpl.java:1412)
	at org.jooq.impl.ParserContext.parseSelect(ParserImpl.java:1191)
	at org.jooq.impl.ParserContext.parseSelect(ParserImpl.java:1186)
	at org.jooq.impl.ParserContext.parseQuery(ParserImpl.java:1041)
	at org.jooq.impl.ParserContext.parseSelect0(ParserImpl.java:844)
	at org.jooq.impl.ParserImpl.parseSelect(ParserImpl.java:697)
	at org.jooq.impl.ParserImpl.parseSelect(ParserImpl.java:692)
	at org.jooq.impl.ParserMetaTest.assertSelectDataTypeEquals(ParserMetaTest.java:494)
	at org.jooq.impl.ParserMetaTest.testParseWithMetaLookupsImplicitSchema(ParserMetaTest.java:1228)
```

The current fix is this:

```java
static final Schema S1a = new SchemaImpl(DSL.name("S1")) {
    @Override
    public List<Table<?>> getTables() {
        return Arrays.asList(S1UpperImplicitSchema.S1_UPPER_IMPLICIT_SCHEMA);
    };
};

static class S1UpperImplicitSchema extends TableImpl<Record> {
    static final S1UpperImplicitSchema S1_UPPER_IMPLICIT_SCHEMA = new S1UpperImplicitSchema();
    final TableField<?, Integer>       COL1                     = createField(DSL.name("COL1"), INTEGER);
    final TableField<?, String>        COL2                     = createField(DSL.name("COL2"), VARCHAR(50));
    final TableField<?, Date>          COL3                     = createField(DSL.name("COL3"), DATE);

    private S1UpperImplicitSchema() {
        super(DSL.name("UPPER"));
    }

    @Override
    public Schema getSchema() {
        return S1a;
    }
}
```

I.e. a `SchemaImpl` instance that references all its tables is required for this lookup to work. This clearly shouldn't be necessary. `TableMetaData` should detect this case, where the tables passed to the constructor have a schema that doesn't back-reference the tables, and patch the schema.

Perhaps, this can be achieved with a call to `snapshot()`?

**Comment by lukaseder** [comment]
Another failing test that doesn't involve parsing:

```java
static class A extends TableImpl/* [java-9] */<Record>/* [/java-9] */ {
    static final A A = new A();
    final TableField<?, Integer> A1 = createField(name("A1"), INTEGER);
    final TableField<?, Integer> A2 = createField(name("A2"), INTEGER);

    A () {
        super(name("A"), schema("S1"));
    }
}

@Test
public void testTableMetaDataImplicitSchema() {
    withMeta(A);

    List<Table<?>> tables = ctx.meta().getTables();
    assertEquals(asList(A), tables);

    Table<?> t = tables.get(0);
    Schema s1 = t.getSchema();
    assertEquals(asList(t), s1.getTables());
    assertEquals("S1", s1.getName());
    assertEquals(asList(s1), s1.getCatalog().getSchemas());
}

private void withMeta(Table<?>... tables) {
    ctx = ctx.configuration().derive((MetaProvider) () -> meta(tables)).dsl();
}
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.8 (#11503)
